client/network/src/service.rs

```diff
@@ -1556,20 +1547,28 @@ where
 					}
 				}
 			},
-			SwarmEvent::ConnectionClosed { peer_id, cause, endpoint, num_established } => {
-				debug!(target: "sub-libp2p", "Libp2p => Disconnected({:?}, {:?})", peer_id, cause);
+			SwarmEvent::ConnectionClosed {
+				connection_id,
+				peer_id,
+				cause,
+				endpoint,
+				num_established,
+			} => {
+				debug!(target: "sub-libp2p", "Libp2p => Disconnected({:?} via {:?}: {:?})", peer_id, connection_id, cause);
 				if let Some(metrics) = self.metrics.as_ref() {
 					let direction = match endpoint {
 						ConnectedPoint::Dialer { .. } => "out",
 						ConnectedPoint::Listener { .. } => "in",
 					};
 					let reason = match cause {
 						Some(ConnectionError::IO(_)) => "transport-error",
+						// Some(ConnectionError::Handler(Either::Left(Either::Left(
```

Comment:
no such variant seems to exist

Reply:
`PingFailure` doesn't exist?

What does `ConnectionError::Handler(Either::Left(Either::Left(Either::Left(Either::Right` contain?

Reply:
```
let reason = match cause {
     |                                        ----- this expression has type `std::option::Option<ConnectionError<either::Either<either::Either<either::Either<either::Either<NotifsHandlerError, either::Either<void::Void, std::io::Error>>, std::io::Error>, void::Void>, void::Void>>>`
```

`std::io::Error`

Reply:
If we are sure that this `io::Error` is handling the case for ping failure, may be we should still report it as `"ping-failure"` and not as default `"protocol-error"`?